src/Vault.sol

```diff
@@ -144,46 +151,64 @@ contract Vault is DefaultAccessControl, IERC721Receiver {
     uint256 public globalStabilisationFeePerUSDSnapshotD = 0;
 
     /// @notice Creates a new contract
-    /// @param admin Protocol admin
     /// @param positionManager_ UniswapV3 position manager
     /// @param factory_ UniswapV3 factory
     /// @param protocolGovernance_ UniswapV3 protocol governance
-    /// @param oracle_ Oracle
     /// @param treasury_ Vault fees treasury
-    /// @param stabilisationFee_ MUSD initial stabilisation fee
     constructor(
-        address admin,
         INonfungiblePositionManager positionManager_,
         IUniswapV3Factory factory_,
         IProtocolGovernance protocolGovernance_,
-        IOracle oracle_,
         address treasury_,
-        uint256 stabilisationFee_
-    ) DefaultAccessControl(admin) {
+        address token_
+    ) {
         if (
             address(positionManager_) == address(0) ||
             address(factory_) == address(0) ||
             address(protocolGovernance_) == address(0) ||
-            address(oracle_) == address(0) ||
-            address(treasury_) == address(0)
+            address(treasury_) == address(0) ||
+            address(token_) == address(0)
         ) {
             revert AddressZero();
         }
 
+        positionManager = positionManager_;
+        factory = factory_;
+        protocolGovernance = protocolGovernance_;
+        treasury = treasury_;
+        token = IMUSD(token_);
+        isInitialized = true;
+    }
+
+    /// @notice Initialized a new contract.
+    /// @param admin Protocol admin
+    /// @param oracle_ Oracle
+    /// @param stabilisationFee_ MUSD initial stabilisation fee
+    function initialize(
+        address admin,
+        IOracle oracle_,
+        uint256 stabilisationFee_
+    ) external {
+        if (isInitialized) {
+            revert Initialized();
+        }
+
+        VaultAccessControl.init(admin);
```

Comment:
Let's prepend all internal functions with an underscore and call them directly, without contract name.
e.g. `_initRoles(admin)` would be fine for me.